### ISSUE TITLE
Backward compatible (1.3.0a0 > 1.2.2 to 1.2.6)

### ISSUE THREAD

**Issue by aereimer** [talmolab/sleap#1217]
"Greetings! I recently experimented with the alpha version 1.3.0a0 of and tested it by opening a project that I initiated in previous versions (1.2.2 to 1.2.6). I added some additional labeled frames, trained the model, and performed successful inferences. However, I encountered issues when attempting to convert the h5 output to slp and then the slp file to csv using a script by @Toshea111 (https://github.com/talmolab/sleap/discussions/1089#discussioncomment-4470031). I suspect that the incompatibility problem might be related to changes in the file structure of the new version's output. Could you confirm if this issue is due to a bug in the newest version or if the file structure of the new version has indeed changed?"

**Comment by talmo** [comment]
Hi @aereimer,

Do you mind sharing the specific error you're getting now? That'll help narrow it down.

Cheers,

Talmo

**Comment by aereimer** [comment]
Hi @talmo, sorry for not pasting the log here before. So, here I tried to open the slp project (created initially in 1.2.2, modified in 1.2.6, and finally modified and saved using 1.3.0a0) using versions < 1.3.0a0 versions. The error below appears specifically when I tried to open the project  using version 1.2.6 ( the same thing happened with other versions, such as 1.2.9). Just one thing, I did not modified the skeleton or added more bodyparts while using the 1.3.0a0 . Thanks for checking this! Parabéns pelo excelente trabalho!

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\commands.py", line 262, in openProject
    self.execute(OpenProject, filename=filename, first_open=first_open)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\commands.py", line 240, in execute
    command().execute(context=self, params=kwargs)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\commands.py", line 137, in execute
    self.do_with_signal(context, params)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\commands.py", line 161, in do_with_signal
    cls.do_action(context, params)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\commands.py", line 606, in do_action
    context.app.loadProjectFile(filename)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\gui\app.py", line 1451, in loadProjectFile
    labels = Labels.load_file(filename, video_search=gui_video_callback)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\dataset.py", line 1943, in load_file
    filename, for_object="labels", video_search=video_search, *args, **kwargs
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\format\main.py", line 113, in read
    return disp.read(filename, *args, **kwargs)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\format\dispatch.py", line 56, in read
    return adaptor.read(file, *args, **kwargs)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\format\hdf5.py", line 139, in read
    labels = cls.read_headers(file, video_search, match_to)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\format\hdf5.py", line 124, in read_headers
    labels = labels_json.LabelsJsonAdaptor.from_json_data(dicts, match_to=match_to)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\io\format\labels_json.py", line 384, in from_json_data
    dicts["skeletons"], List[Skeleton]
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\cattr\converters.py", line 201, in structure
    return self._structure_func.dispatch(cl)(obj, cl)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\cattr\converters.py", line 336, in _structure_list
    for e in obj
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\cattr\converters.py", line 336, in <listcomp>
    for e in obj
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\skeleton.py", line 248, in <lambda>
    Skeleton, lambda x, cls: Skeleton.from_dict(x, idx_to_node)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\skeleton.py", line 931, in from_dict
    return Skeleton.from_json(json.dumps(d), node_to_idx)
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\sleap\skeleton.py", line 1030, in from_json
    graph = json_graph.node_link_graph(jsonpickle.decode(json_str))
  File "C:\ProgramData\Anaconda3\envs\sleap_develop\lib\site-packages\networkx\readwrite\json_graph\node_link.py", line 166, in node_link_graph
    for d in data["nodes"]:
KeyError: 'nodes'

**Comment by Toshea111** [comment]
Dear aereimer,

The issue might be that you are trying to convert '.slp' files, when the notebook is designed for converting '.h5' files to '.csv'.

Try directly inputting the '.h5' output, rather than converting it into '.slp' format first.

Best wishes,

Thomas

**Comment by aereimer** [comment]
Dear @Toshea111,

Thanks for pointing this out. I believe I’ve tried first the slp file because the colab file originally was pointing to a slp file ( talk_title_slide@13150-14500.mp4.predictions.slp ). But I also tried the h5 file (exported from SLEAP v1.3.0a0 ) and it did not work, the error message is below - that's also the reason I've asked about changes in the output file in this newer version.

Thanks once again for you input.
Best,
Adriano

`KeyError                                  Traceback (most recent call last)
[<ipython-input-6-8f2541ce267e>](https://localhost:8080/#) in <module>
----> 1 labels = sio.load_slp("labels_sleap13.v002.000_R14_EPM_session_2_20211116_45_adjusted.analysis.h5")
      2 # labels = sleap.load_file("sleap1.3")
      3 labels

4 frames
h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

[/usr/local/lib/python3.9/dist-packages/h5py/_hl/group.py](https://localhost:8080/#) in __getitem__(self, name)
    286                 raise ValueError("Invalid HDF5 object reference")
    287         else:
--> 288             oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
    289 
    290         otype = h5i.get_type(oid)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5o.pyx in h5py.h5o.open()

KeyError: "Unable to open object (object 'tracks_json' doesn't exist)"`

**Comment by Toshea111** [comment]
Dear aereimer,

Looking at the link, I think that is an older version of the notebook.

There is an updated version here:

https://github.com/Toshea111/sleap/blob/develop/docs/notebooks/Convert_HDF5_to_CSV_updated.ipynb

Let me know if the issue persists.

Best wishes,

Thomas

**Comment by roomrys** [comment]
Hi @aereimer,

I can sheepishly confirm that the file structure has changed. We [added two new attributes in the `Skeleton` class](https://github.com/talmolab/sleap/pull/1122/files) called `description` and `preview_image`. Consequently, while you are able to open older files on v1.3.0a0, files from this version are unable to be opened on previous versions - which is not a problem if working on a single SLEAP version. However, I can see how this could easily become extremely annoying if working on multiple workstations which all have different versions of SLEAP installed.

The field we added is currently not used except for provided template skeletons, so we could selectively write the field depending on whether the skeleton is a template (manually specified via API - only by those looking to add to our collection of template skeletons). I opened PR #1222 to rectify this.

Thanks,
Liezl

**Comment by aereimer** [comment]
Dear @Toshea111,

I have tried the new script that you shared, and it works great! It is compatible with all SLEAP versions that I have tested. I made minor changes to the script to suit my project's requirements (as it has a single track; I still wanted to get the score for each body part/frame), but apart from that, it is perfect. Thank you once again for sharing the notebook!

Best,
Adriano

**Comment by Toshea111** [comment]
Dear Adriano,

Excellent, and good to hear that it works with the latest version.

@roomrys I am happy to incorporate this into a future analysis notebook, if that would be useful?

Best wishes,

Thomas

**Comment by roomrys** [comment]
Hi @Toshea111,

Yes! This notebook is gaining a lot of traction. We could even convert your notebook into an adaptor for using in the GUI/API/CLI instead of having it as a separate entity. If you start a PR, we can collaborate (not sure if you would be able work on a talmolab/sleap branch, but I can pop to your fork to get the basic structure set-up). We would need to
1. Add a adaptor file to the [format folder](https://github.com/talmolab/sleap/tree/develop/sleap/io/format) which implements a subclass of the `Adaptor` class
https://github.com/talmolab/sleap/blob/4bd017150acad47e7586c891fe9855eef0b4c48f/sleap/io/format/adaptor.py#L20-L28

2. Connect the adaptor to the GUI (various places)
3. Connect the adaptor to the `sleap-convert` command
https://github.com/talmolab/sleap/blob/4bd017150acad47e7586c891fe9855eef0b4c48f/sleap/io/convert.py#L100-L105

I think copying some of the notebook code into the different methods in the adaptor would be a great start for integration. Let me know what you think. Totally understand if this is more than what you wanted to bite off, and I can take over whatever you'd rather not tackle at the moment.

Thanks,
Liezl

**Comment by Toshea111** [comment]
Dear Liezl,

I would be happy to collaborate, although I fear that my expertise is limited.

I made the fork 'Toshea111/sleap' a while back, and will invite you as a collaborator.

The notebook is currently located in:

https://github.com/Toshea111/sleap/tree/develop/docs/notebooks

The impetus for the notebook is some downstream software that I am working on, thus further integration would be useful.

Best wishes,

Thomas

**Comment by roomrys** [comment]
This fix is now available in the latest release v1.3.0. To install:
```
conda create -y -n sleap -c sleap -c nvidia -c conda-forge sleap=1.3.0
```

**Comment by Toshea111** [comment]
Dear Liezl,

Much appreciated.

Best wishes,

Thomas
